examples/direct-upload-to-gcs/appengine/main.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2/google"
+	iam "google.golang.org/api/iam/v1"
+)
+
+var (
+	// iamService is a client for calling the signBlob API.
+	iamService *iam.Service
+
+	// serviceAccountName represents Service Account Name.
+	// See more details: https://cloud.google.com/iam/docs/service-accounts
+	serviceAccountName string
+
+	// serviceAccountID follows the below format.
+	// "projects/%s/serviceAccounts/%s"
+	serviceAccountID string
+
+	// uploadableBucket is the destination bucket.
+	// All users will upload files directly to this bucket by using generated Signed URL.
+	uploadableBucket string
+)
+
+func init() {
+	cred, err := google.DefaultClient(context.Background(), iam.CloudPlatformScope)
+	if err != nil {
+		log.Fatal(err)
+	}
+	iamService, err = iam.New(cred)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	uploadableBucket = os.Getenv("UPLOADABLE_BUCKET")
+	serviceAccountName = os.Getenv("SERVICE_ACCOUNT")
+	serviceAccountID = fmt.Sprintf(
+		"projects/%s/serviceAccounts/%s",
+		os.Getenv("GOOGLE_CLOUD_PROJECT"),
+		serviceAccountName,
+	)
+}
+
+func signHandler(w http.ResponseWriter, r *http.Request) {
+	// Accepts only POST method.
+	// Otherwise, this handler returns 405.
+	if m := r.Method; m != "POST" {
```

Comment:
There's no need to copy r.Method to a new variable just to test it.

```go
if r.Method != "POST" {
```

Reply:
`m` is referred from https://github.com/GoogleCloudPlatform/professional-services/pull/214/files#diff-8f26846eb7bceff3a368b00752552a15R73

Reply:
While you could possibly make an argument for not calling the `FormValue()` method below more than once, there's no reason to copy r.Method to a new variable just to test it (and then print it). You're not changing the value in any way so this only make the code (a VERY SMALL amount) slower by having to copy a string, and make the code very slightly more complex.

Yes, I'm being pedantic about it, but readability is not improved by having single-use variables created when this would do just fine:

```go
if r.Method != "POST" {
  w.Header().Set("Allow", "POST")
  http.Error(w, "Only POST is supported", http.StatusMethodNotAllowed)
  return
}
```

You don't even need to send back the original method in the error message to the client. The client should already know what method was used in the request.

Also, the server must send back an `Allow` header, not `Allows` as you have it here.

Reply:
First off, thanks for pointing the typo out, it's very helpful.

---

About the copy to new variable, it's a copy cost to the variable you care about, but it is actually a code that can be commonly found in the development of Go itself.
https://github.com/golang/go/blob/92e6cf4c0f9bf2e1ecd57f22a430d96020cbdd69/src/net/http/server.go#L949-L950
https://github.com/golang/go/blob/92e6cf4c0f9bf2e1ecd57f22a430d96020cbdd69/src/net/http/server.go#L1797

Of course, this should have been used for a lot of testings, like `want, got := tt.want, foo`.
I don't say you're wrong, but I don't think the opinion what you said is treated as a best practice in all aspects of Go. 
also ref: https://groups.google.com/forum/#!topic/golang-nuts/ZRKSJ3GPkLw


However, I will follow you, as I'm not an owner and not familiar with this repository's coding rule.

Reply:
Except the case that requiring calling method multiple times (e.g. `FormValue()`), I have just removed variable assignment for this.
https://github.com/GoogleCloudPlatform/professional-services/pull/214/commits/54ace8d2ba16da50f31fefedb2c1b0c61f25938e

Reply:
Thank you for the consideration.

I was curious about the actual performance difference so I threw together a benchmark:
```go
func BenchmarkWithCopy(b *testing.B) {
  r := &http.Request{Method: "GET"}
  for i := 0; i < b.N; i++ {
    if m := r.Method; m != "POST" {
      fmt.Sprintf("%s", m)
    }
  }
}

func BenchmarkWithoutCopy(b *testing.B) {
  r := &http.Request{Method: "GET"}
  for i := 0; i < b.N; i++ {
    if r.Method != "POST" {
      fmt.Sprintf("%s", r.Method)
    }
  }
}
```

```shell
$ go test -bench=. tmp_test.go
goos: linux
goarch: amd64
BenchmarkWithCopy-8             10000000               151 ns/op
BenchmarkWithoutCopy-8          10000000               145 ns/op
PASS
ok      command-line-arguments  4.484s
```

The actual cost of copying is *tiny* but it's there. In this case it's statistically insignificant.

Reply:
Coincidentally, I also tried the exact same benchmark :)
Yep, the cost of copying definitely exists, but it's very tiny because of actual memory of data is not copied, only just [its pointer and length of data](https://github.com/golang/go/blob/92e6cf4c0f9bf2e1ecd57f22a430d96020cbdd69/src/runtime/string.go#L218). 
I wish it could be CoW, but Go's string is immutable.

Btw, could you review other points?